kura/test/org.eclipse.kura.cloudconnection.sparkplug.mqtt.provider.test/src/test/java/org/eclipse/kura/cloudconnection/sparkplug/mqtt/endpoint/test/SubscriptionsMapTest.java

```diff
@@ -0,0 +1,196 @@
+        thenMatchingListenersContains("t1", 0, this.l2);
+        thenReturnedTopicsToUnsubscribeIsEmpty();
+    }
+
+    @Test
+    public void removeAllShouldReturnTopicsToUnsubscribe() {
+        givenAdd("t1", 0, this.l1);
+        givenAdd("t1", 0, this.l2);
+        givenRemove(this.l1);
+
+        whenRemove(this.l2);
+
+        thenSubscriptionRecordsNotContains(new SubscriptionRecord("t1", 0));
+        thenMatchingListenersNotContains("t1", 0, this.l2);
+        thenReturnedTopicsToUnsubscribeContains("t1");
+    }
+
+    @Test
+    public void shouldMatchSingleLevelWildCard1() {
+        givenAdd("A/B", 0, this.l1);
+        givenAdd("A/+/C", 0, this.l2);
+        givenAdd("A/B/C", 0, this.l3);
+
+        whenGetListenersMatching("A/B/C", 0);
+
+        thenReturnedMatchingListenersNotContains(this.l1);
+        thenReturnedMatchingListenersContains(this.l2);
+        thenReturnedMatchingListenersContains(this.l3);
+    }
+
+    @Test
+    public void shouldMatchSingleLevelWildCard2() {
+        givenAdd("A/B", 0, this.l1);
+        givenAdd("A/+", 0, this.l2);
+        givenAdd("A/B/C", 0, this.l3);
+
+        whenGetListenersMatching("A/B", 0);
+
+        thenReturnedMatchingListenersContains(this.l1);
+        thenReturnedMatchingListenersContains(this.l2);
+        thenReturnedMatchingListenersNotContains(this.l3);
+    }
+
+    @Test
+    public void shouldMatchMultiLevelWildCard() {
+        givenAdd("A/B", 0, this.l1);
+        givenAdd("A/#", 0, this.l2);
+        givenAdd("A/B/C", 0, this.l3);
+
+        whenGetListenersMatching("A/B/C", 0);
+
+        thenReturnedMatchingListenersNotContains(this.l1);
+        thenReturnedMatchingListenersContains(this.l2);
+        thenReturnedMatchingListenersContains(this.l3);
+    }
+
+    @Test
+    public void shouldMatchWithCorrectQos() {
+        givenAdd("A/B", 1, this.l1);
+        givenAdd("A/#", 0, this.l2);
+        givenAdd("A/B/C", 1, this.l3);
+
+        whenGetListenersMatching("A/B/C", 0);
+
+        thenReturnedMatchingListenersNotContains(this.l1);
+        thenReturnedMatchingListenersContains(this.l2);
+        thenReturnedMatchingListenersNotContains(this.l3);
+    }
+
+    @Test
+    public void shouldMatchWithMaximumSubscribedQos() {
+        givenAdd("A/B/C", 0, this.l1);
+
+        whenGetListenersMatching("A/B/C", 1);
+
+        thenReturnedMatchingListenersContains(this.l1);
+    }
+
+    /*
+     * Steps
+     */
+
+    /*
+     * Given
+     */
+
+    private void givenAdd(String topicFilter, int qos, CloudSubscriberListener listener) {
+        this.subscriptionsMap.add(topicFilter, qos, listener);
+    }
+
+    private void givenRemove(CloudSubscriberListener listener) {
+        this.returnedTopicsToUnsubscribe.addAll(this.subscriptionsMap.remove(listener));
+    }
+
+    /*
+     * When
+     */
+
+    private void whenRemove(CloudSubscriberListener listener) {
+        givenRemove(listener);
+    }
```

Comment:
This make me feel like the "given" should be a "when" really... If you need to remove multiple listeners for your test consider having a "when" method that iterates over a list.